extensions/amp-story/1.0/amp-story-page.js

```diff
@@ -1310,27 +1313,27 @@ export class AmpStoryPage extends AMP.BaseElement {
 
     const hasAudioPromise = hasAudioElements
       ? Promise.resolve(true)
-      : this.hasVideoWithAudio_();
+      : this.hasLoadedVideoWith_(
+          (video) =>
+            !video.hasAttribute('noaudio') &&
+            parseFloat(video.getAttribute('volume')) !== 0
+        );
 
     hasAudioPromise.then((hasAudio) =>
       this.storeService_.dispatch(Action.TOGGLE_PAGE_HAS_AUDIO, hasAudio)
     );
   }
 
   /**
-   * Checks if the page has any videos with audio.
+   * Checks if the page has any videos that meet the requirement after receiving the video cache response.
+   * @param {function<!Element, boolean>} func
    * @return {!Promise<boolean>}
    * @private
    */
-  hasVideoWithAudio_() {
+  hasLoadedVideoWith_(func) {
```

Comment:
As a first read, I really disagree with this code, can you correct me if I'm wrong?

`amp-video.buildCallback()` will only resolve when the video cache call resolved and is processed (sources are appended, `audio` attr is set, and `track` is appended).

`waitForElementsWithUnresolvedAudio` retrieves the component impl through `getImpl()` which sounds unnecessary, it should rely on a post buildCallback signal. Same to wait for tracks to be appended.

If you don't want to write new code you can rely on `waitForPlaybackMediaLayout_` but technically you could even do sooner than that.
On top of being too complicated for no good reason, and building on top of it, this code is also wrong and doesn't rely on the media selectors that we built to look within friendly iframes (ads).

I'd really want this code to re-use the existing helpers and not build a custom broken solution.

Reply:
I agree that waiting for `waitForPlaybackMediaLayout` is too long, we could wait for `whenUpgradedToCustomElement(mediaEl).then((el) => el.signals().whenSignal(BUILD))` but it's longer than `getImpl()` and does the same thing. I think renaming the method we use to `waitForVideosWithCachedSources` makes it more explicit and then we can do the logic to find if any of the videos have tracks